charts/rancher-windows-gmsa/1.0.0/templates/_helpers.tpl

```diff
@@ -0,0 +1,35 @@
+# Rancher
+
+{{- define "system_default_registry" -}}
+{{- if .Values.global.cattle.systemDefaultRegistry -}}
+{{- printf "%s/" .Values.global.cattle.systemDefaultRegistry -}}
+{{- end -}}
+{{- end -}}
+
+{{/* Create chart name and version as used by the chart label. */}}
+{{- define "gmsa.chartref" -}}
+chart: {{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}
+{{- end }}
+
+{{/* Determine apiVersion for cert-manager */}}
+{{- define "cert-manager.apiversion" -}}
+  {{- $certmanagerVer :=  split "." .Values.certificates.certManager.version -}}
+  {{- if or (.Capabilities.APIVersions.Has "cert-manager.io/v1") (and (gt (len $certmanagerVer._0) 0) (eq (int $certmanagerVer._0) 1) (ge (int $certmanagerVer._1) 0)) }}
+apiVersion: cert-manager.io/v1
+  {{- else if or (.Capabilities.APIVersions.Has "cert-manager.io/v1beta1") (and (gt (len $certmanagerVer._0) 0) (eq (int $certmanagerVer._0) 0) (ge (int $certmanagerVer._1) 16)) }}
+apiVersion: cert-manager.io/v1beta1
+  {{- else if or (.Capabilities.APIVersions.Has "cert-manager.io/v1alpha2") (and (gt (len $certmanagerVer._0) 0) (eq (int $certmanagerVer._0) 0) (ge (int $certmanagerVer._1) 11)) }}
+apiVersion: cert-manager.io/v1alpha2
+  {{- else if or (.Capabilities.APIVersions.Has "certmanager.k8s.io/v1alpha1") (and (gt (len $certmanagerVer._0) 0) (eq (int $certmanagerVer._0) 0) (lt (int $certmanagerVer._1) 11)) }}
+apiVersion: cert-manager.io/v1alpha1
+  {{- else }}
+apiVersion: cert-manager.io/v1
+  {{- end }}
+{{- end }}
+
+{{- define "certificates.cabundle"}}
+{{- if gt (len (lookup "rbac.authorization.k8s.io/v1" "ClusterRole" "" "")) 0 -}}
+{{- (lookup "v1" "Secret" "" "{{ .Release.Namespace }}" "{{ .Values.certificates.secretName }}").data.ca.crt }}
+{{- end -}}
+{{- end }}

```

Comment:
I don't think this works, this needs to be tested. The lookup function only takes 4 arguments, you've listed 5 here.

Here's a fixed version you can test with checks on the contents of the secret as well:

```yaml
{{- define "certificates.cabundle"}}
{{- $secret := (lookup "v1" "Secret" .Release.Namespace .Values.certificates.secretName) -}}
{{- if lt (len $secret) 1 -}}
{{- required (printf "CA Bundle  secret '%s' in namespace '%s' must exist" .Values.certificates.secretName .Release.Namespace) "" -}}
{{- else -}}
{{- required (printf "CA Bundle secret '%s' in namespace '%s' is empty" .Values.certificates.secretName .Release.Namespace) $secret.data -}}
{{- if or (not (hasKey $secret.data "ca.crt")) (not (hasKey $secret.data "tls.crt")) (not (hasKey $secret.data "tls.key")) -}}
{{- required (printf "CA Bundle secret '%s' in namespace '%s' must contain ca.crt, tls.key, and tls.cert; found the following keys in the secret: %s" .Values.certificates.secretName .Release.Namespace $secret.data) "" -}}
{{- end -}}
{{- end -}}
{{- get $secret.data "ca.crt" }}
{{- end }}
```

Reply:
Ran test cases locally:

```bash
arvindiyengar: ~/Rancher/charts/src/github.com/rancher/charts 
$ helm upgrade --install rancher-windows-gmsa packages/rancher-windows-gmsa/charts
Error: UPGRADE FAILED: execution error at (rancher-windows-gmsa/templates/_helpers.tpl:34:4): CA Bundle secret 'gmsa-server-cert' in namespace 'default' must exist
arvindiyengar: ~/Rancher/charts/src/github.com/rancher/charts 
$ k create secret generic gmsa-server-cert                                        
secret/gmsa-server-cert created
arvindiyengar: ~/Rancher/charts/src/github.com/rancher/charts 
$ helm upgrade --install rancher-windows-gmsa packages/rancher-windows-gmsa/charts
Error: UPGRADE FAILED: execution error at (rancher-windows-gmsa/templates/_helpers.tpl:37:4): CA Bundle secret 'gmsa-server-cert' in namespace 'default' is empty
arvindiyengar: ~/Rancher/charts/src/github.com/rancher/charts 
$ k edit secret gmsa-server-cert                                                  
secret/gmsa-server-cert edited
arvindiyengar: ~/Rancher/charts/src/github.com/rancher/charts 
$ helm upgrade --install rancher-windows-gmsa packages/rancher-windows-gmsa/charts
Error: UPGRADE FAILED: execution error at (rancher-windows-gmsa/templates/_helpers.tpl:40:4): CA Bundle secret 'gmsa-server-cert' in namespace 'default' must contain ca.crt, tls.key, and tls.cert; found the following keys in the secret: map[hi:aGVsbG8K]
arvindiyengar: ~/Rancher/charts/src/github.com/rancher/charts 
$ k edit secret gmsa-server-cert                                                  
secret/gmsa-server-cert edited
arvindiyengar: ~/Rancher/charts/src/github.com/rancher/charts 
$ helm upgrade --install rancher-windows-gmsa packages/rancher-windows-gmsa/charts
Error: UPGRADE FAILED: execution error at (rancher-windows-gmsa/templates/_helpers.tpl:40:4): CA Bundle secret 'gmsa-server-cert' in namespace 'default' must contain ca.crt, tls.key, and tls.cert; found the following keys in the secret: map[ca.crt:aGVsbG8K]
arvindiyengar: ~/Rancher/charts/src/github.com/rancher/charts 
$ k edit secret gmsa-server-cert                                                  
secret/gmsa-server-cert edited
arvindiyengar: ~/Rancher/charts/src/github.com/rancher/charts 
$ helm upgrade --install rancher-windows-gmsa packages/rancher-windows-gmsa/charts
Release "rancher-windows-gmsa" has been upgraded. Happy Helming!
NAME: rancher-windows-gmsa
LAST DEPLOYED: Thu Feb 10 14:29:06 2022
NAMESPACE: default
STATUS: deployed
REVISION: 5
TEST SUITE: None
```